### ISSUE TITLE
Disable caching / refresh data on history operations

### ISSUE THREAD

**Issue by mattkingshott** [inertiajs/inertia#565]
Hello.

The docs mention the following about "remembering state"

```
When navigating browser history, Inertia restores pages using prop data cached in history state.
```

Is there any way to turn off this functionality, but which I mean, have it so that when you click back, it fetches fresh data from the server and renders the component using the revised data?

I ask, because I'm having trouble with (for example, notifications re-appearing on the page when I click back). The simplest solution seems to be to configure behaviour so that "back" (and I guess "forward" too, or really any sort history operation) triggers Inertia to fetch fresh data. It doesn't need to be a full page reload, just the data then re-render the Vue component.

Thanks!

P.S. I could also imagine there being value by adding a no caching option e.g. if the contents of a particular page were sensitive and you wanted to force a data refresh / re-render.

**EDIT**

I've come up with this in `app.js` before we initialise Vue, what do you think @reinink? 

It creates a pop state handler before Inertia is set up, this handler then prevents event propagation, which prevents Inertia doing its thing. Then, within the handler we do a reload and assign the page:

```js
window.addEventListener('popstate', (event) => {
    event.stopImmediatePropagation();

    Inertia.reload({
        preserveState  : false,
        preserveScroll : false,
        replace        : true,
        onSuccess      : (page) => Inertia.setPage(page),
        onError        : () => window.location.href = event.state.url,
    });
});
```

I had an earlier, much more complicated option, which I've since removed since the above achieves the same thing and is so much more cleaner.

**Comment by mbryne** [comment]
Great snippet there Matt, 

That did solve the issue for me but found that it loses the ability to move forward through the history,

It seems like for toasts and notifications etc it would be better to filter our certain values being stored in the history, similar this issue: https://github.com/inertiajs/inertia/issues/56

I went down the road of trying to filter that state but ran out of time so will keep an eye on this issue.

**Comment by mattkingshott** [comment]
Thanks @mbryne.

I've actually just updated it with `replace : true`, which fixed some issues for me. You might want to see if it solves the issue for you.

**Comment by freakzlike** [comment]
I faced the same issue when using back after some updates have been performed. The code snippet works for me but I am using typescript and receiving an error with `Inertia.setPage(page)` because `setPage` is `protected`.

Is the `onSuccess` callback necessary? It also works without it.

I think there should be at least a global config option to force a refetch on every history navigation.

**Comment by rovansteen** [comment]
This snippet worked a bit more reliable for me:

```js
let stale = false;

window.addEventListener('popstate', () => {
  stale = true;
});

Inertia.on('navigate', (event) => {
  const page = event.detail.page;
  if (stale) {
    Inertia.get(page.url, {}, { replace: true, preserveState: false });
  }
  stale = false;
});
```

**Comment by mattkingshott** [comment]
@freakzlike I encountered errors without using `onSuccess`, but I can't remember what they were I'm afraid.

@rovansteen When you say "more reliable", what errors did you encounter that made you switch? I'm curious as I may use your snippet instead.

**Comment by brian-dillingham-odin** [comment]
Would love if this was a config option.

**Comment by margalit** [comment]
I stumbled on this issue as my flash messages were re-trigging between back/forward navigations since the flash message was being retrieved from the cache in my react app.

The solution was just to reload the flash props after I show the message:
```javascript
const useFlashMessage = () => {
  const { flash } = usePage().props;
  const { showNotification } = useNotifications();
  useEffect(() => {
    if (flash.message) {
      showNotification({
        message: flash.message,
      });
      Inertia.reload({ only: ["flash"] });
    }
  }, [flash.message]);
};
```

**Comment by claudiodekker** [comment]
Hi all,

While I appreciate your workarounds for this, we will not be implementing this feature or support for it, as this is not how the history state in browsers were designed to work. Even _if_ we were to build an implementation, we would be relying on non-standard/unsupported behavior that might change or break without notice as browsers ship updates.

For more information on _why_ this is unsupported behavior, you can read my comment that I posted on it here: https://github.com/inertiajs/inertia/issues/247#issuecomment-704375830

**Comment by janis-verins** [comment]
To not reload all page props, one solution would be to include, which data should be reloaded for a certain page when `popstate` event is triggered.
For example, in the Controller:
`    public function indexUser(Request $request): Response
    {
        return Inertia::render('User/Communication', ['revalidate_on_popstate_data' => ['data']]);
    }`

And in app.js:
`window.addEventListener('popstate', (event) => {
    event.stopImmediatePropagation();

    const { state: { props: { revalidate_on_popstate_data = [] } } } = event;

    Inertia.reload({
        only: revalidate_on_popstate_data,
        preserveState: false,
        preserveScroll: true,
        replace: true
    });
});`

**Comment by thetanaz** [comment]
Call me a newb but I do not understand how this is the default behavior on history events and there is no obvious way outside of custom hooks to change it. There are so many situations in which users interract with their browser history and they need fresh data when navigating back and forth. Let's say for example you have a chat functionality in your app, and then you go to check something and you go back to the chat you had open - then you're seeing old data. Even if you use the useRemember() to remember the messages that have been exchanged while you were there, you still won't see any new messages that you received while you navigated away. I can name at least 5 more scenarios like this, I mean there's gotta be a way to have this functionality without "hacking" things together.

**Comment by dkornilove** [comment]
Yeah, it is one of the painfull things when I try to integrate redux with inertia

**Comment by jacekobst1** [comment]
Here's the solution for Inertia v2.0. Just add this event listener in the app.ts:

```
    import { router } from '@inertiajs/vue3';

    window.addEventListener('popstate', (event) => {
        event.stopImmediatePropagation();
        router.get(location.href, {}, { replace: true, preserveState: false });
    });
```

**Comment by aneesdev** [comment]
For Inertia v2:

```javascript
window.addEventListener("popstate", (event) => {
  event.stopImmediatePropagation();

  router.visit(window.location.href, {
    preserveState: false,
    preserveScroll: false,
    replace: true,
  });
});
```

**Comment by Petar-CV** [comment]
Leaving this in case anybody stumbles onto this issue. My problem with calling `router.reload({ only: ['flash'] });` whenever toast message is shown is that it'll call the API once again to only refresh the toast prop. I don't like these duplicate requests.

My use case is that I wanted to refetch data whenever `popstate` happened. However, I have a provider that automatically shows toast messages whenever it detects `flash` property inside the `usePage` hook. As you might've guessed, these flash messages reappeared on `popstate` because Inertia cached the previous page and restored it before I had the opportunity to reset the state.
The following code works perfectly in my case since I don't need to cache data on backward navigation. This avoids the duplicate request whenever toast message is shown.

```typescript
window.addEventListener("popstate", (event) => {
    event.stopImmediatePropagation();

    router.reload({
        except: ["flash"],
    });
});
````

**Comment by gurgeous** [comment]
Just want to thank everyone for the workarounds. We love inertia! This is a bit of a footgun and it would be nice to have an note in the docs or perhaps a workaround. Like a list of pages that get tossed on back button, perhaps. We are doing that internally for now.

**Comment by clarkmcc** [comment]
@claudiodekker Is it possible to disable caching entirely? This is a footgun that's getting me on almost every turn:
* I use the `HandleInertiaResponse.php` middleware to keep track of whether the sidebar is expanded or collapsed. The problem is when I navigate to a page that's cached, it returns the cached sidebar state, not the latest sidebar state.
* I have an impersonation mode but when I navigate backwards after impersonating someone, I get the cached version of the page as my original user, rather than as my impersonated user.
* When I perform an action that decrements my notifications counter, but then navigate backwards, the counter appears incremented again.

I was able to solve that last problem by calling `Inertia::clearHistory()`, but it did not resolve my impersonation issue for some reason, and I don't want to have to stick this basically anywhere where I'm performing an application state change operation.

In summary, it would be great if I could simply disable the cache. It's creating more challenges than its worth.

**Comment by clarkmcc** [comment]
Bumping this

> @claudiodekker Is it possible to disable caching entirely? This is a footgun that's getting me on almost every turn:
> * I use the `HandleInertiaResponse.php` middleware to keep track of whether the sidebar is expanded or collapsed. The problem is when I navigate to a page that's cached, it returns the cached sidebar state, not the latest sidebar state.
> * I have an impersonation mode but when I navigate backwards after impersonating someone, I get the cached version of the page as my original user, rather than as my impersonated user.
> * When I perform an action that decrements my notifications counter, but then navigate backwards, the counter appears incremented again.
> 
> I was able to solve that last problem by calling `Inertia::clearHistory()`, but it did not resolve my impersonation issue for some reason, and I don't want to have to stick this basically anywhere where I'm performing an application state change operation.
> 
> In summary, it would be great if I could simply disable the cache. It's creating more challenges than its worth.

**Comment by Danita** [comment]
Bumping this. Any way to disable caching?

**Comment by reinink** [comment]
@Danita Yes, with history encryption. See here: https://inertiajs.com/docs/v2/security/history-encryption

**Comment by EriksonEnterprises** [comment]
> @Danita Yes, with history encryption. See here: https://inertiajs.com/docs/v2/security/history-encryption

This should honestly be on by default and then just let users call clearHistory if desired, or disable it if desired

But yes this works! Thank you! Just call clearHistory on the page/controller you want to reset prior history from (as it has both controller and client side clearing), and it works :)

It's also in v3 :)
